File: jtransc-game/src/jtransc/game/util/Signal.java
package jtransc.game.util;

import java.util.ArrayList;

public class Signal<T> {
	ArrayList<SignalHandler<T>> handlers = null;

	public void add(SignalHandler<T> value) {
        if (handlers == null) {
            handlers = new ArrayList<SignalHandler<T>>();
        }
		handlers.add(value);
	}

	public void remove(SignalHandler<T> value) {
        if (handlers != null) {
            handlers.remove(value);
        }
	}

	public void dispatch(T value) {
        if (handlers != null) {
            for (SignalHandler<T> handler : handlers) {
                handler.handle(value);
            }
        }
	}
}

File: jtransc-utils/src/com/jtransc/async/signal.kt
package com.jtransc.async

import com.jtransc.lang.Disposable

class Signal<T : Any> {
	private val callbacks = arrayListOf<(value: T) -> Unit>()

	private class SignalDisposable<T : Any>(val signal: Signal<T>, val callback: (value: T) -> Unit) : Disposable {
		override fun dispose() {
			signal.remove(callback)
		}
	}

	fun removeAll() {
		callbacks.clear()
	}

	private fun remove(callback: (value: T) -> Unit) {
		callbacks.remove(callback)
	}

	fun add(callback: (value: T) -> Unit): Disposable {
		callbacks.add(callback)
		return SignalDisposable(this, callback)
	}

	fun once(callback: (value: T) -> Unit): Disposable {
		var handler: (T) -> Unit = {}
		handler = { value: T ->
			remove(handler)
			callback(value)
		}
		add(handler)
		return SignalDisposable(this, handler)
	}

	fun dispatch(value: T) {
		for (c in callbacks.toList()) c(value)
	}

	operator fun invoke(value: T) = dispatch(value)
	operator fun invoke(callback: (value: T) -> Unit) = add(callback)
}

fun Signal<Unit>.dispatch() {
	return this.dispatch(Unit)
}

fun <T : Any> Signal<T>.pipeTo(that: Signal<T>) {
	this.add { that.dispatch(it) }
}

fun <T : Any> Signal<T>.waitOneAsync(): Promise<T> {
	val deferred = Promise.Deferred<T>()
	this.once { deferred.resolve(it) }
	return deferred.promise
}
